File: lib/services/email_service.dart
import 'dart:convert';
import 'dart:math';
import 'package:http/http.dart' as http;

class EmailService {
  // SMTP2GO Configuration
  static const String smtp2goApiKey = 'api-5C5BC5F363E84C4D9C9302E57F873F57'; // Replace with your actual API key
  static const String smtp2goEndpoint = 'https://api.smtp2go.com/v3/email/send';
  static const String fromEmail = '121deepak2104@sjcem.edu.in'; // Replace with your verified sender email
  static const String fromName = 'MicroNest';
  
  // Generate a random 6-digit OTP
  static String generateOTP() {
    Random random = Random();
    return (100000 + random.nextInt(900000)).toString();
  }
  
  // Send email OTP
  static Future<Map<String, dynamic>> sendEmailOTP(String toEmail, String otp) async {
    try {
      final emailData = {
        'api_key': smtp2goApiKey,
        'to': [toEmail],
        'sender': fromEmail,
        'subject': 'MicroNest - Email Verification OTP',
        'html_body': _generateOTPEmailHTML(otp),
        'text_body': _generateOTPEmailText(otp),
        'custom_headers': [
          {
            'header': 'Reply-To',
            'value': fromEmail
          }
        ]
      };

      final response = await http.post(
        Uri.parse(smtp2goEndpoint),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode(emailData),
      );

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        if (responseData['data'] != null && responseData['data']['succeeded'] == 1) {
          return {
            'success': true,
            'message': 'OTP sent successfully',
            'otp': otp,
          };
        } else {
          return {
            'success': false,
            'message': 'Failed to send email: ${responseData['data']?['error'] ?? 'Unknown error'}',
          };
        }
      } else {
        return {
          'success': false,
          'message': 'HTTP Error: ${response.statusCode}',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Network error: ${e.toString()}',
      };
    }
  }

  // Send welcome email
  static Future<Map<String, dynamic>> sendWelcomeEmail(String toEmail, String fullName) async {
    try {
      final emailData = {
        'api_key': smtp2goApiKey,
        'to': [toEmail],
        'sender': fromEmail,
        'subject': 'Welcome to MicroNest!',
        'html_body': _generateWelcomeEmailHTML(fullName),
        'text_body': _generateWelcomeEmailText(fullName),
        'custom_headers': [
          {
            'header': 'Reply-To',
            'value': fromEmail
          }
        ]
      };

      final response = await http.post(
        Uri.parse(smtp2goEndpoint),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode(emailData),
      );

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        if (responseData['data'] != null && responseData['data']['succeeded'] == 1) {
          return {
            'success': true,
            'message': 'Welcome email sent successfully',
          };
        } else {
          return {
            'success': false,
            'message': 'Failed to send welcome email: ${responseData['data']?['error'] ?? 'Unknown error'}',
          };
        }
      } else {
        return {
          'success': false,
          'message': 'HTTP Error: ${response.statusCode}',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Network error: ${e.toString()}',
      };
    }
  }

  // Send password reset email
  static Future<Map<String, dynamic>> sendPasswordResetEmail(String toEmail, String resetToken) async {
    try {
      final resetLink = 'https://micronest.devloperwala.in/reset-password?token=$resetToken';
      
      final emailData = {
        'api_key': smtp2goApiKey,
        'to': [toEmail],
        'sender': fromEmail,
        'subject': 'MicroNest - Password Reset Request',
        'html_body': _generatePasswordResetEmailHTML(resetLink),
        'text_body': _generatePasswordResetEmailText(resetLink),
        'custom_headers': [
          {
            'header': 'Reply-To',
            'value': fromEmail
          }
        ]
      };

      final response = await http.post(
        Uri.parse(smtp2goEndpoint),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode(emailData),
      );

      if (response.statusCode == 200) {
        final responseData = jsonDecode(response.body);
        if (responseData['data'] != null && responseData['data']['succeeded'] == 1) {
          return {
            'success': true,
            'message': 'Password reset email sent successfully',
          };
        } else {
          return {
            'success': false,
            'message': 'Failed to send password reset email: ${responseData['data']?['error'] ?? 'Unknown error'}',
          };
        }
      } else {
        return {
          'success': false,
          'message': 'HTTP Error: ${response.statusCode}',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Network error: ${e.toString()}',
      };
    }
  }

  // Generate HTML email template for OTP
  static String _generateOTPEmailHTML(String otp) {
    return '''
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Email Verification - MicroNest</title>
        <style>
            body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
            .container { max-width: 600px; margin: 0 auto; padding: 20px; }
            .header { background: linear-gradient(135deg, #52B788, #40916C); color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }
            .content { background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px; }
            .otp-box { background: #52B788; color: white; padding: 20px; text-align: center; border-radius: 8px; margin: 20px 0; font-size: 24px; font-weight: bold; letter-spacing: 5px; }
            .footer { text-align: center; margin-top: 30px; color: #666; font-size: 14px; }
        </style>
    </head>
    <body>
        <div class="container">
            <div class="header">
                <h1>μN</h1>
                <h2>Email Verification</h2>
            </div>
            <div class="content">
                <p>Hello!</p>
                <p>Thank you for signing up with MicroNest. To complete your registration, please use the verification code below:</p>
                
                <div class="otp-box">$otp</div>
                
                <p>This code will expire in 10 minutes for security reasons.</p>
                <p>If you didn't request this verification, please ignore this email.</p>
                
                <p>Best regards,<br>The MicroNest Team</p>
            </div>
            <div class="footer">
                <p>© 2024 MicroNest. All rights reserved.</p>
            </div>
        </div>
    </body>
    </html>
    ''';
  }

  // Generate text email template for OTP
  static String _generateOTPEmailText(String otp) {
    return '''
Email Verification - MicroNest

Hello!

Thank you for signing up with MicroNest. To complete your registration, please use the verification code below:

$otp

This code will expire in 10 minutes for security reasons.

If you didn't request this verification, please ignore this email.

Best regards,
The MicroNest Team

© 2024 MicroNest. All rights reserved.
    ''';
  }

  // Generate HTML email template for welcome email
  static String _generateWelcomeEmailHTML(String fullName) {
    return '''
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Welcome to MicroNest!</title>
        <style>
            body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
            .container { max-width: 600px; margin: 0 auto; padding: 20px; }
            .header { background: linear-gradient(135deg, #52B788, #40916C); color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }
            .content { background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px; }
            .feature { background: white; padding: 20px; margin: 15px 0; border-radius: 8px; border-left: 4px solid #52B788; }
            .footer { text-align: center; margin-top: 30px; color: #666; font-size: 14px; }
        </style>
    </head>
    <body>
        <div class="container">
            <div class="header">
                <h1>μN</h1>
                <h2>Welcome to MicroNest!</h2>
            </div>
            <div class="content">
                <p>Hello $fullName!</p>
                <p>Welcome to MicroNest! Your account has been successfully created and verified.</p>
                
                <div class="feature">
                    <h3>🚀 What's Next?</h3>
                    <p>• Complete your profile setup<br>
                    • Join or create savings groups<br>
                    • Start building your financial future</p>
                </div>
                
                <div class="feature">
                    <h3>🔒 Your Account is Secure</h3>
                    <p>• Email verified successfully<br>
                    • Password encrypted and secure<br>
                    • Two-factor authentication available</p>
                </div>
                
                <p>Ready to get started? Log in to your account and explore the features!</p>
                
                <p>Best regards,<br>The MicroNest Team</p>
            </div>
            <div class="footer">
                <p>© 2024 MicroNest. All rights reserved.</p>
            </div>
        </div>
    </body>
    </html>
    ''';
  }

  // Generate text email template for welcome email
  static String _generateWelcomeEmailText(String fullName) {
    return '''
Welcome to MicroNest!

Hello $fullName!

Welcome to MicroNest! Your account has been successfully created and verified.

What's Next?
• Complete your profile setup
• Join or create savings groups
• Start building your financial future

Your Account is Secure
• Email verified successfully
• Password encrypted and secure
• Two-factor authentication available

Ready to get started? Log in to your account and explore the features!

Best regards,
The MicroNest Team

© 2024 MicroNest. All rights reserved.
    ''';
  }

  // Generate HTML email template for password reset
  static String _generatePasswordResetEmailHTML(String resetLink) {
    return '''
    <!DOCTYPE html>
    <html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Password Reset - MicroNest</title>
        <style>
            body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
            .container { max-width: 600px; margin: 0 auto; padding: 20px; }
            .header { background: linear-gradient(135deg, #52B788, #40916C); color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }
            .content { background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px; }
            .reset-button { background: #52B788; color: white; padding: 15px 30px; text-decoration: none; border-radius: 8px; display: inline-block; margin: 20px 0; }
            .footer { text-align: center; margin-top: 30px; color: #666; font-size: 14px; }
        </style>
    </head>
    <body>
        <div class="container">
            <div class="header">
                <h1>μN</h1>
                <h2>Password Reset Request</h2>
            </div>
            <div class="content">
                <p>Hello!</p>
                <p>We received a request to reset your MicroNest account password. Click the button below to create a new password:</p>
                
                <a href="$resetLink" class="reset-button">Reset Password</a>
                
                <p>If the button doesn't work, copy and paste this link into your browser:</p>
                <p>$resetLink</p>
                
                <p>This link will expire in 1 hour for security reasons.</p>
                <p>If you didn't request a password reset, please ignore this email and your password will remain unchanged.</p>
                
                <p>Best regards,<br>The MicroNest Team</p>
            </div>
            <div class="footer">
                <p>© 2024 MicroNest. All rights reserved.</p>
            </div>
        </div>
    </body>
    </html>
    ''';
  }

  // Generate text email template for password reset
  static String _generatePasswordResetEmailText(String resetLink) {
    return '''
Password Reset - MicroNest

Hello!

We received a request to reset your MicroNest account password. Click the link below to create a new password:

$resetLink

This link will expire in 1 hour for security reasons.

If you didn't request a password reset, please ignore this email and your password will remain unchanged.

Best regards,
The MicroNest Team

© 2024 MicroNest. All rights reserved.
    ''';
  }
} 

File: backend/models/EmailService.php
<?php

class EmailService {
    private $smtp2goApiKey;
    private $smtp2goApiUrl;
    private $fromEmail;
    private $fromName;
    
    public function __construct() {
        // SMTP2GO Configuration - Update these with your actual credentials
        $this->smtp2goApiKey = 'api-XXXXXXXXXXXX'; // Replace with your actual SMTP2GO API key
        $this->smtp2goApiUrl = 'https://api.smtp2go.com/v3/email/send';
        $this->fromEmail = 'noreply@micronest.com';
        $this->fromName = 'MicroNest';
        
        // For development/testing, you can also use environment variables
        if (getenv('SMTP2GO_API_KEY')) {
            $this->smtp2goApiKey = getenv('SMTP2GO_API_KEY');
        }
        if (getenv('SMTP2GO_FROM_EMAIL')) {
            $this->fromEmail = getenv('SMTP2GO_FROM_EMAIL');
        }
    }
    
    /**
     * Send email using SMTP2GO
     */
    public function sendEmail($to, $subject, $htmlBody, $textBody = null) {
        try {
            if (empty($this->smtp2goApiKey) || $this->smtp2goApiKey === 'YOUR_SMTP2GO_API_KEY') {
                // For testing purposes, log the email instead of sending
                error_log("Email would be sent to: $to");
                error_log("Subject: $subject");
                error_log("HTML Body: $htmlBody");
                return [
                    'success' => true,
                    'message' => 'Email logged (SMTP2GO not configured)',
                    'to' => $to,
                    'subject' => $subject
                ];
            }
            
            $emailData = [
                'api_key' => $this->smtp2goApiKey,
                'to' => [$to],
                'sender' => $this->fromEmail,
                'subject' => $subject,
                'html_body' => $htmlBody,
                'text_body' => $textBody ?: strip_tags($htmlBody)
            ];
            
            $ch = curl_init();
            curl_setopt($ch, CURLOPT_URL, $this->smtp2goApiUrl);
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($emailData));
            curl_setopt($ch, CURLOPT_HTTPHEADER, [
                'Content-Type: application/json',
                'Accept: application/json'
            ]);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_TIMEOUT, 30);
            curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, true);
            
            $response = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $error = curl_error($ch);
            curl_close($ch);
            
            if ($error) {
                throw new Exception("cURL Error: $error");
            }
            
            if ($httpCode !== 200) {
                throw new Exception("SMTP2GO API returned HTTP $httpCode: $response");
            }
            
            $result = json_decode($response, true);
            
            if (isset($result['data']['succeeded']) && $result['data']['succeeded'] > 0) {
                return [
                    'success' => true,
                    'message' => 'Email sent successfully',
                    'to' => $to,
                    'subject' => $subject
                ];
            } else {
                throw new Exception("SMTP2GO API error: " . json_encode($result));
            }
            
        } catch (Exception $e) {
            error_log("EmailService Error: " . $e->getMessage());
            return [
                'success' => false,
                'message' => 'Failed to send email: ' . $e->getMessage(),
                'to' => $to,
                'subject' => $subject
            ];
        }
    }
    
    /**
     * Send OTP email
     */
    public function sendOtpEmail($email, $fullName, $otp) {
        $subject = 'MicroNest - Email Verification OTP';
        $htmlBody = $this->generateOtpEmailHTML($fullName, $otp);
        $textBody = $this->generateOtpEmailText($fullName, $otp);
        
        return $this->sendEmail($email, $subject, $htmlBody, $textBody);
    }
    
    /**
     * Generate HTML email template for OTP
     */
    private function generateOtpEmailHTML($fullName, $otp) {
        return "
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset='UTF-8'>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <title>Email Verification OTP</title>
            <style>
                body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
                .container { max-width: 600px; margin: 0 auto; padding: 20px; }
                .header { background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }
                .content { background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px; }
                .otp-box { background: #fff; border: 2px solid #667eea; border-radius: 8px; padding: 20px; text-align: center; margin: 20px 0; }
                .otp-code { font-size: 32px; font-weight: bold; color: #667eea; letter-spacing: 5px; }
                .footer { text-align: center; margin-top: 30px; color: #666; font-size: 14px; }
            </style>
        </head>
        <body>
            <div class='container'>
                <div class='header'>
                    <h1>MicroNest</h1>
                    <p>Email Verification</p>
                </div>
                <div class='content'>
                    <h2>Hello $fullName!</h2>
                    <p>Thank you for signing up with MicroNest. To complete your registration, please use the following verification code:</p>
                    
                    <div class='otp-box'>
                        <div class='otp-code'>$otp</div>
                        <p><strong>This code will expire in 10 minutes.</strong></p>
                    </div>
                    
                    <p>If you didn't request this verification, please ignore this email.</p>
                    
                    <p>Best regards,<br>The MicroNest Team</p>
                </div>
                <div class='footer'>
                    <p>This is an automated message, please do not reply to this email.</p>
                </div>
            </div>
        </body>
        </html>
        ";
    }
    
    /**
     * Generate plain text email template for OTP
     */
    private function generateOtpEmailText($fullName, $otp) {
        return "
MicroNest - Email Verification OTP

Hello $fullName!

Thank you for signing up with MicroNest. To complete your registration, please use the following verification code:

OTP: $otp

This code will expire in 10 minutes.

If you didn't request this verification, please ignore this email.

Best regards,
The MicroNest Team

---
This is an automated message, please do not reply to this email.
        ";
    }
    
    /**
     * Send welcome email
     */
    public function sendWelcomeEmail($email, $fullName) {
        $subject = 'Welcome to MicroNest!';
        $htmlBody = $this->generateWelcomeEmailHTML($fullName);
        $textBody = $this->generateWelcomeEmailText($fullName);
        
        return $this->sendEmail($email, $subject, $htmlBody, $textBody);
    }
    
    /**
     * Generate HTML email template for welcome email
     */
    private function generateWelcomeEmailHTML($fullName) {
        return "
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset='UTF-8'>
            <meta name='viewport' content='width=device-width, initial-scale=1.0'>
            <title>Welcome to MicroNest</title>
            <style>
                body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
                .container { max-width: 600px; margin: 0 auto; padding: 20px; }
                .header { background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 30px; text-align: center; border-radius: 10px 10px 0 0; }
                .content { background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px; }
                .welcome-box { background: #fff; border: 2px solid #667eea; border-radius: 8px; padding: 20px; text-align: center; margin: 20px 0; }
                .footer { text-align: center; margin-top: 30px; color: #666; font-size: 14px; }
            </style>
        </head>
        <body>
            <div class='container'>
                <div class='header'>
                    <h1>MicroNest</h1>
                    <p>Welcome Aboard!</p>
                </div>
                <div class='content'>
                    <h2>Welcome to MicroNest, $fullName!</h2>
                    
                    <div class='welcome-box'>
                        <p>🎉 Your account has been successfully created and verified!</p>
                        <p>You can now access all the features of MicroNest and start building your financial future.</p>
                    </div>
                    
                    <p>Here's what you can do next:</p>
                    <ul>
                        <li>Complete your profile</li>
                        <li>Join or create savings groups</li>
                        <li>Start contributing to build your savings</li>
                        <li>Apply for loans when needed</li>
                    </ul>
                    
                    <p>If you have any questions, feel free to contact our support team.</p>
                    
                    <p>Best regards,<br>The MicroNest Team</p>
                </div>
                <div class='footer'>
                    <p>This is an automated message, please do not reply to this email.</p>
                </div>
            </div>
        </body>
        </html>
        ";
    }
    
    /**
     * Generate plain text email template for welcome email
     */
    private function generateWelcomeEmailText($fullName) {
        return "
Welcome to MicroNest!

Hello $fullName!

🎉 Your account has been successfully created and verified!

You can now access all the features of MicroNest and start building your financial future.

Here's what you can do next:
- Complete your profile
- Join or create savings groups
- Start contributing to build your savings
- Apply for loans when needed

If you have any questions, feel free to contact our support team.

Best regards,
The MicroNest Team

---
This is an automated message, please do not reply to this email.
        ";
    }
    
    /**
     * Set SMTP2GO API key
     */
    public function setApiKey($apiKey) {
        $this->smtp2goApiKey = $apiKey;
    }
    
    /**
     * Set sender email
     */
    public function setFromEmail($email) {
        $this->fromEmail = $email;
    }
    
    /**
     * Set sender name
     */
    public function setFromName($name) {
        $this->fromName = $name;
    }
}
?> 